tunnel/tunnel.go

```diff
@@ -1175,124 +1173,127 @@ func (c *Client) NewSession(target Target) (_ io.ReadWriteCloser, err error) {
 		defer func() {
 			<-c.block
 		}()
-		// wrap critical section in a function to allow use of defer
-		setup := func() error {
-			c.cmu.Lock()
-			defer c.cmu.Unlock()
-			stream, err := c.tc.Register(ctx, c.cc.Opts...)
-			if err != nil {
-				return fmt.Errorf("start: failed to create register stream: %v", err)
-			}
-			c.rs = &regSafeStream{regStream: stream}
-			p, ok := peer.FromContext(stream.Context())
-			if !ok {
-				return errors.New("no peer from stream context")
-			}
-			c.addr = p.Addr
-
-			for target := range c.targets {
-				c.NewTarget(target)
-			}
+		errCh := make(chan error, 1)
 
-			for typ := range c.peerTypeTargets {
-				c.Subscribe(typ)
-			}
-
-			return nil
-		}
-
-		if err := setup(); err != nil {
-			return err
-		}
-		c.Registered = true
-		return c.start(ctx)
-	default:
-		return errors.New("client is already running")
-	}
-}
-
-// start handles received register stream requests.
-func (c *Client) start(ctx context.Context) error {
-	errCh := make(chan error, 1)
-
-	go func() {
-		for {
-			reg, err := c.rs.Recv()
-			if err != nil {
-				select {
-				case errCh <- err:
-				default:
-				}
-				return
-			}
-
-			switch reg.Registration.(type) {
-			case *tpb.RegisterOp_Session:
-				session := reg.GetSession()
-				if !session.GetAccept() {
+		go func() {
+			for {
+				reg, err := c.rs.Recv()
+				if err != nil {
 					select {
-					case errCh <- fmt.Errorf("connection %d not accepted by server", session.GetTag()):
+					case errCh <- err:
 					default:
 					}
 					return
 				}
-				tag := session.Tag
-				tID := session.GetTargetId()
-				tType := session.GetTargetType()
-				go func() {
-					if err := c.streamHandler(ctx, tag, Target{ID: tID, Type: tType}); err != nil {
+
+				switch reg.Registration.(type) {
+				case *tpb.RegisterOp_Session:
+					session := reg.GetSession()
+					if !session.GetAccept() {
 						select {
-						case errCh <- err:
+						case errCh <- fmt.Errorf("connection %d not accepted by server", session.GetTag()):
 						default:
 						}
+						return
 					}
-				}()
-			case *tpb.RegisterOp_Target:
-				target := reg.GetTarget()
-				switch op := target.GetOp(); op {
-				case tpb.Target_ADD:
-					if err := c.addPeerTarget(target); err != nil {
-						errCh <- fmt.Errorf("failed to add peer target: %v", err)
-					}
-				case tpb.Target_REMOVE:
-					if err := c.deletePeerTarget(target); err != nil {
-						errCh <- fmt.Errorf("failed to delete peer target: %v", err)
-					}
-				default:
-					// The rest are acks.
-					if !target.GetAccept() {
-						select {
-						case errCh <- fmt.Errorf("target registration (%s, %s) not accepted by server", target.TargetId, target.TargetType):
-						default:
+					tag := session.Tag
+					tID := session.GetTargetId()
+					tType := session.GetTargetType()
+					go func() {
+						if err := c.streamHandler(ctx, tag, Target{ID: tID, Type: tType}); err != nil {
+							select {
+							case errCh <- err:
```

Comment:
If you write to errCh here then Start will return, but the goroutine started on line 1224 will keep running, ie. it will be leaked and any other streamHandler goroutines will be leaked. On the other hand if the goroutine started on line 1224 hits an error then Start will return, but any of these streamHandler goroutines will keep running and be leaked.

This can be refactored to be more simple by removing the outer goroutine. For example,

```
ctx, cancel := context.WithCancel(ctx)
defer cancel() // Tell the streamHandler goroutines to exit when this function exits
for {
    reg, err := c.s.Recv()
    if err != nil {
        return err
    }
    switch reg.Registration.(type) {
    case *tpb.RegisterOp_Session:
       ...
       go func() { 
           if err := c.streamHandler(ctx, ...); err != nil {
               log.Printf("streamHandler error: %v", err)
           }
       }()
```
This simplifies things by removing one goroutine and the error channel. The code can now just return errors. If an error is encountered Start will return and cancel the context passed to any streamHandler goroutines, which should cause them to end. This means nothing is leaked. A streamHandler error doesn't cause this whole function to exit anymore. I'm not sure if that is desirable or not?

Reply:
Thanks for the comment. I removed the goroutine. It looks nicer.
For the streamHandler, think we probably still need it to propagate the error for downstream. For example, in conn.go  Listener.Accept will check if an error or a connection is received in order to proceeds.